File: btap-necb/test/test_loads_apply_loads.rb
require_relative 'test_helper'

# P3 gate (standalone half): assign_space_types on-ramp + apply_loads golden
# assertions (people/equipment densities, DSOA rescale, infiltration, schedule
# wiring, thermostats, skips, coverage emission).
class TestApplyLoads < Minitest::Test
  include FixtureHelper

  MAP_TYPES = {
    'office' => ['Space Function', 'Office enclosed > 25 m2'],
    'corridor' => ['Space Function', 'Corridor/Transition area other-sch-A'],
    'dining' => ['Space Function', 'Dining area - family dining']
  }.freeze

  def mapped_model
    model = load_fixture
    spaces = model.getSpaces.sort_by(&:nameString)
    map = {}
    spaces.each_with_index do |space, index|
      map[space.nameString] = index.zero? ? MAP_TYPES['corridor'] : MAP_TYPES['office']
    end
    [model, map]
  end

  def applied_model
    model, map = mapped_model
    audit = BtapNECB::AuditLog.new
    BtapNECB::Loads.assign_space_types(model, map, vintage: '2020', audit: audit)
    BtapNECB::Loads.apply_loads(model, vintage: '2020', audit: audit)
    [model, audit]
  end

  def office_space_type(model)
    model.getSpaceTypes.find { |st| st.nameString.include?('Office enclosed > 25 m2') }
  end

  def test_assign_space_types_on_ramp
    model, map = mapped_model
    audit = BtapNECB::AuditLog.new
    BtapNECB::Loads.assign_space_types(model, map, vintage: '2020', audit: audit)
    decision = audit.entries.find { |e| e[:action] == 'NECB space types assigned' }
    assert_equal 2, decision[:inputs][:space_types_created], 'one SpaceType per distinct pair'
    model.getSpaces.each do |space|
      assert space.spaceType.is_initialized
      assert space.spaceType.get.standardsBuildingType.is_initialized
    end
    assert_raises(ArgumentError) do
      BtapNECB::Loads.assign_space_types(model, { model.getSpaces.first.nameString => %w[Nope Nada] })
    end
  end

  def test_people_and_equipment_golden
    model, = applied_model
    office = office_space_type(model)
    record = BtapNECB::Loads::SpaceTypes.record(
      building_type: 'Space Function', space_type: 'Office enclosed > 25 m2')

    people = office.people.first
    expected_density = OpenStudio.convert(record['occupancy_per_area'].to_f / 1000, 'people/ft^2', 'people/m^2').get
    assert_in_delta expected_density, people.peopleDefinition.peopleperSpaceFloorArea.get, 1e-9
    assert_in_delta 0.3, people.peopleDefinition.fractionRadiant, 1e-9
    assert people.clothingInsulationSchedule.is_initialized, 'comfort schedules wired'

    equip = office.electricEquipment.first
    expected_epd = OpenStudio.convert(record['electric_equipment_per_area'].to_f, 'W/ft^2', 'W/m^2').get
    assert_in_delta expected_epd, equip.electricEquipmentDefinition.wattsperSpaceFloorArea.get, 1e-9

    assert_empty office.lights.to_a, 'NO Lights objects — lighting gem territory'
  end

  def test_ventilation_rescale_and_stash
    model, = applied_model
    office = office_space_type(model)
    record = BtapNECB::Loads::SpaceTypes.record(
      building_type: 'Space Function', space_type: 'Office enclosed > 25 m2')
    dsoa = office.designSpecificationOutdoorAir.get

    assert_equal 'Sum', dsoa.outdoorAirMethod
    per_person = record['ventilation_per_person'].to_f
    unless per_person.zero?
      expected = per_person * record['ventilation_occupancy_rate_people_per_1000ft2'].to_f /
                 record['occupancy_per_area'].to_f
      expected_si = OpenStudio.convert(expected, 'ft^3/min*person', 'm^3/s*person').get
      assert_in_delta expected_si, dsoa.outdoorAirFlowperPerson, 1e-9, 'per-person RESCALE applied'
    end
    stash = dsoa.additionalProperties
    assert stash.getFeatureAsDouble('Ref OA per person').is_initialized, 'source values stashed'
    assert stash.getFeatureAsString('Ref standard').is_initialized
  end

  def test_schedules_thermostats_and_infiltration
    model, = applied_model
    office = office_space_type(model)
    record = BtapNECB::Loads::SpaceTypes.record(
      building_type: 'Space Function', space_type: 'Office enclosed > 25 m2')

    schedule_set = office.defaultScheduleSet.get
    assert_equal record['occupancy_schedule'],
                 schedule_set.numberofPeopleSchedule.get.nameString
    assert schedule_set.electricEquipmentSchedule.is_initialized
    assert schedule_set.lightingSchedule.empty?, 'no lighting schedule wired'

    thermostat = model.getThermostatSetpointDualSetpoints.find { |t| t.nameString == "#{office.nameString} Thermostat" }
    refute_nil thermostat
    assert_equal record['heating_setpoint_schedule'],
                 thermostat.heatingSetpointTemperatureSchedule.get.nameString

    unless record['infiltration_per_exterior_area'].to_f.zero?
      infiltration = office.spaceInfiltrationDesignFlowRates.first
      expected = OpenStudio.convert(record['infiltration_per_exterior_area'].to_f, 'ft^3/min*ft^2', 'm^3/s*m^2').get
      assert_in_delta expected, infiltration.flowperExteriorSurfaceArea.get, 1e-9
    end
  end

  def test_skips_and_coverage
    model, map = mapped_model
    plenum = OpenStudio::Model::SpaceType.new(model)
    plenum.setName('Attic plenum')
    audit = BtapNECB::AuditLog.new
    BtapNECB::Loads.assign_space_types(model, map, vintage: '2020', audit: audit)
    BtapNECB::Loads.apply_loads(model, vintage: '2020', audit: audit)

    assert(audit.entries.any? { |e| e[:action].include?('plenum space type skipped') })
    coverage = audit.entries.select { |e| e[:step] == :coverage }
    # 8.4.3.2. is declared PER SENTENCE since the coverage-depth pass: (1)/(2)
    # partial (cross-gem schedule delegations, illuminance), (3) modeller scope.
    # 4 other 8.4.3.x + 8.4.2.7 (internal loads slice) + 8.4.3.6 outdoor air.
    assert_equal 9, coverage.size, 'all declared entries accounted'
    partial = coverage.find { |e| e[:article] == '8.4.3.2.(1)' }
    assert_equal :warning, partial[:level], 'partial status WARNS (lighting+SHW schedule delegations)'
    assert_match(/openstudio-lighting/, partial[:action])
    semi = coverage.find { |e| e[:article] == '8.4.3.2.(3)' }
    assert_equal :info, semi[:level], '(3) semi-heated set-point is a modeller input, not a warning'
    decisions = audit.entries.select { |e| e[:article].to_s.include?('8.4.3.2') && e[:level] == :decision }
    refute_empty decisions
  end

  def test_2025_citation_prefix_flows_to_audit
    model, map = mapped_model
    audit = BtapNECB::AuditLog.new
    BtapNECB::Loads.assign_space_types(model, map, vintage: '2025', audit: audit)
    BtapNECB::Loads.apply_loads(model, vintage: '2025', audit: audit)
    office = office_space_type(model)
    refute_nil office.people.first, '2025 aliases the 2020 data'
  end
end

File: python/tests/necb/test_loads_apply_loads.py
"""P3 gate (standalone half): assign_space_types on-ramp + apply_loads golden
assertions (people/equipment densities, DSOA rescale, infiltration, schedule
wiring, thermostats, skips, coverage emission).

Port of btap-necb/test/test_loads_apply_loads.rb. The parity half
(test_loads_apply_parity.rb) needs the live Ruby oracle; its Leg-C equivalent is
test_oracle_goldens_loads.py.
"""

from __future__ import annotations

import unittest

from tests.support import load_fixture, needs_sdk

MAP_TYPES = {
    'office': ['Space Function', 'Office enclosed > 25 m2'],
    'corridor': ['Space Function', 'Corridor/Transition area other-sch-A'],
    'dining': ['Space Function', 'Dining area - family dining'],
}


@needs_sdk
class TestApplyLoads(unittest.TestCase):
    def mapped_model(self):
        from btap._compat import sorted_by_name
        model = load_fixture()
        spaces = sorted_by_name(model.getSpaces())
        map_ = {}
        for index, space in enumerate(spaces):
            map_[space.nameString()] = (MAP_TYPES['corridor'] if index == 0
                                        else MAP_TYPES['office'])
        return model, map_

    def applied_model(self):
        from btap.audit import AuditLog
        from btap.necb import loads
        model, map_ = self.mapped_model()
        audit = AuditLog()
        loads.assign_space_types(model, map_, vintage='2020', audit=audit)
        loads.apply_loads(model, vintage='2020', audit=audit)
        return model, audit

    def office_space_type(self, model):
        return next((st for st in model.getSpaceTypes()
                     if 'Office enclosed > 25 m2' in st.nameString()), None)

    def test_assign_space_types_on_ramp(self):
        from btap.audit import AuditLog
        from btap.necb import loads
        model, map_ = self.mapped_model()
        audit = AuditLog()
        loads.assign_space_types(model, map_, vintage='2020', audit=audit)
        decision = next(e for e in audit.entries
                        if e['action'] == 'NECB space types assigned')
        self.assertEqual(2, decision['inputs']['space_types_created'],
                         'one SpaceType per distinct pair')
        for space in model.getSpaces():
            self.assertTrue(space.spaceType().is_initialized())
            self.assertTrue(space.spaceType().get().standardsBuildingType().is_initialized())
        with self.assertRaises(ValueError):
            loads.assign_space_types(
                model, {model.getSpaces()[0].nameString(): ['Nope', 'Nada']})

    def test_people_and_equipment_golden(self):
        import openstudio

        from btap.necb import loads
        model, _ = self.applied_model()
        office = self.office_space_type(model)
        record = loads.SpaceTypes.record(building_type='Space Function',
                                         space_type='Office enclosed > 25 m2')

        people = office.people()[0]
        expected_density = openstudio.convert(
            float(record['occupancy_per_area']) / 1000, 'people/ft^2', 'people/m^2').get()
        self.assertAlmostEqual(expected_density,
                               people.peopleDefinition().peopleperSpaceFloorArea().get(),
                               delta=1e-9)
        self.assertAlmostEqual(0.3, people.peopleDefinition().fractionRadiant(), delta=1e-9)
        self.assertTrue(people.clothingInsulationSchedule().is_initialized(),
                        'comfort schedules wired')

        equip = office.electricEquipment()[0]
        expected_epd = openstudio.convert(
            float(record['electric_equipment_per_area']), 'W/ft^2', 'W/m^2').get()
        self.assertAlmostEqual(
            expected_epd, equip.electricEquipmentDefinition().wattsperSpaceFloorArea().get(),
            delta=1e-9)

        self.assertEqual([], list(office.lights()),
                         'NO Lights objects — lighting domain territory')

    def test_ventilation_rescale_and_stash(self):
        import openstudio

        from btap.necb import loads
        model, _ = self.applied_model()
        office = self.office_space_type(model)
        record = loads.SpaceTypes.record(building_type='Space Function',
                                         space_type='Office enclosed > 25 m2')
        dsoa = office.designSpecificationOutdoorAir().get()

        self.assertEqual('Sum', dsoa.outdoorAirMethod())
        per_person = float(record['ventilation_per_person'])
        if per_person != 0:
            expected = (per_person
                        * float(record['ventilation_occupancy_rate_people_per_1000ft2'])
                        / float(record['occupancy_per_area']))
            expected_si = openstudio.convert(
                expected, 'ft^3/min*person', 'm^3/s*person').get()
            self.assertAlmostEqual(expected_si, dsoa.outdoorAirFlowperPerson(), delta=1e-9,
                                   msg='per-person RESCALE applied')
        stash = dsoa.additionalProperties()
        self.assertTrue(stash.getFeatureAsDouble('Ref OA per person').is_initialized(),
                        'source values stashed')
        self.assertTrue(stash.getFeatureAsString('Ref standard').is_initialized())

    def test_schedules_thermostats_and_infiltration(self):
        import openstudio

        from btap.necb import loads
        model, _ = self.applied_model()
        office = self.office_space_type(model)
        record = loads.SpaceTypes.record(building_type='Space Function',
                                         space_type='Office enclosed > 25 m2')

        schedule_set = office.defaultScheduleSet().get()
        self.assertEqual(record['occupancy_schedule'],
                         schedule_set.numberofPeopleSchedule().get().nameString())
        self.assertTrue(schedule_set.electricEquipmentSchedule().is_initialized())
        self.assertFalse(schedule_set.lightingSchedule().is_initialized(),
                         'no lighting schedule wired')

        wanted = f"{office.nameString()} Thermostat"
        thermostat = next((t for t in model.getThermostatSetpointDualSetpoints()
                           if t.nameString() == wanted), None)
        self.assertIsNotNone(thermostat)
        self.assertEqual(record['heating_setpoint_schedule'],
                         thermostat.heatingSetpointTemperatureSchedule().get().nameString())

        if float(record['infiltration_per_exterior_area']) != 0:
            infiltration = office.spaceInfiltrationDesignFlowRates()[0]
            expected = openstudio.convert(float(record['infiltration_per_exterior_area']),
                                          'ft^3/min*ft^2', 'm^3/s*m^2').get()
            self.assertAlmostEqual(expected,
                                   infiltration.flowperExteriorSurfaceArea().get(),
                                   delta=1e-9)

    def test_skips_and_coverage(self):
        import openstudio

        from btap.audit import AuditLog
        from btap.necb import loads
        model, map_ = self.mapped_model()
        plenum = openstudio.model.SpaceType(model)
        plenum.setName('Attic plenum')
        audit = AuditLog()
        loads.assign_space_types(model, map_, vintage='2020', audit=audit)
        loads.apply_loads(model, vintage='2020', audit=audit)

        self.assertTrue(any('plenum space type skipped' in e['action']
                            for e in audit.entries))
        coverage = [e for e in audit.entries if e['step'] == 'coverage']
        # 8.4.3.2. is declared PER SENTENCE since the coverage-depth pass: (1)/(2)
        # partial (cross-gem schedule delegations, illuminance), (3) modeller scope.
        # 4 other 8.4.3.x + 8.4.2.7 (internal loads slice) + 8.4.3.6 outdoor air.
        self.assertEqual(9, len(coverage), 'all declared entries accounted')
        partial = next(e for e in coverage if e.get('article') == '8.4.3.2.(1)')
        self.assertEqual('warning', partial['level'],
                         'partial status WARNS (lighting+SHW schedule delegations)')
        self.assertRegex(partial['action'], 'lighting')
        semi = next(e for e in coverage if e.get('article') == '8.4.3.2.(3)')
        self.assertEqual('info', semi['level'],
                         '(3) semi-heated set-point is a modeller input, not a warning')
        decisions = [e for e in audit.entries
                     if '8.4.3.2' in str(e.get('article') or '')
                     and e['level'] == 'decision']
        self.assertNotEqual([], decisions)

    def test_2025_citation_prefix_flows_to_audit(self):
        from btap.audit import AuditLog
        from btap.necb import loads
        model, map_ = self.mapped_model()
        audit = AuditLog()
        loads.assign_space_types(model, map_, vintage='2025', audit=audit)
        loads.apply_loads(model, vintage='2025', audit=audit)
        office = self.office_space_type(model)
        self.assertTrue(len(office.people()) > 0, '2025 aliases the 2020 data')


if __name__ == '__main__':
    unittest.main()
